wrappers/kmip/kmip.go

```diff
@@ -0,0 +1,512 @@
+	return wrapping.WrapperTypeKmip, nil
+}
+
+// KeyId returns the last known key id
+func (k *Wrapper) KeyId(_ context.Context) (string, error) {
+	return k.keyId, nil
+}
+
+// Encrypt is used to encrypt the master key using the the KMIP.
+// This returns the ciphertext, and/or any errors from this
+// call. This should be called after the KMS client has been instantiated.
+func (k *Wrapper) Encrypt(ctx context.Context, plaintext []byte, opt ...wrapping.Option) (*wrapping.BlobInfo, error) {
+	if plaintext == nil {
+		return nil, fmt.Errorf("given plaintext for encryption is nil")
+	}
+
+	ciphertext, nonce, err := k.encryptKMIP(ctx, plaintext)
+	if err != nil {
+		return nil, fmt.Errorf("error encrypting data: %w", err)
+	}
+
+	ret := &wrapping.BlobInfo{
+		Ciphertext: ciphertext,
+		Iv:         nonce,
+		KeyInfo: &wrapping.KeyInfo{
+			KeyId:     k.keyId, // This is the decryption key ID (ie: the AES key or the RSA private key)
+			Mechanism: k.mechanismID,
+		},
+	}
+
+	return ret, nil
+}
+
+// Decrypt is used to decrypt the ciphertext. This should be called after Init.
+func (k *Wrapper) Decrypt(ctx context.Context, in *wrapping.BlobInfo, opt ...wrapping.Option) ([]byte, error) {
+	if in == nil {
+		return nil, fmt.Errorf("given input for decryption is nil")
+	}
+
+	mId := in.KeyInfo.Mechanism
+	if mId >= uint64(len(mechanisms)) {
+		return nil, fmt.Errorf("invalid encryption mechanism ID %q", mId)
+	}
+	cparams, ok := cryptoParamsPreset[mechanisms[mId]]
+	if !ok {
+		return nil, fmt.Errorf("mechanism ID does not match any KMIP cryptographic parameter")
+	}
+
+	plaintext, err := k.decryptKMIP(ctx, in.KeyInfo.KeyId, cparams, in.Iv, in.Ciphertext)
+	if err != nil {
+		return nil, fmt.Errorf("error decrypting data encryption key: %w", err)
+	}
+
+	return plaintext, nil
+}
+
+// encryptKMIP uses the KMIP encrypt operation to encrypt the DEK.
+func (kms *Wrapper) encryptKMIP(ctx context.Context, plainDEK []byte) ([]byte, []byte, error) {
+	var iv []byte
+	var err error
+	if ivl := kms.cryptoParams.IVLength; ivl > 0 {
+		iv, err = uuid.GenerateRandomBytes(int(ivl))
+		if err != nil {
+			return nil, nil, err
+		}
+	}
+
+	client, err := kms.connect(ctx)
+	if err != nil {
+		return nil, nil, err
+	}
+	defer client.Close()
+
+	keyId := kms.keyId
+
+	// When using RSA, some KMIP servers may accept encryption using the private key ID, some others only accept the public key ID.
+	// Let's play it safe and always use the public key for encryption.
+	// The public key id is looked-up during initialization. It's empty when running in symmetric mode.
+	if kms.pubKeyId != "" {
+		keyId = kms.pubKeyId
+	}
+
+	encryptRespPayload, err := client.Encrypt(keyId).
+		WithIvCounterNonce(iv).
+		WithCryptographicParameters(kms.cryptoParams).
+		Data(plainDEK).
+		ExecContext(ctx)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// Append the encryption tag (if any) to the ciphertext. Depending on the KMIP version and the implementation,
```

Comment:
Just checking; does this mean we'll double-append in the case of it being in the tag _and_ already appended? Or does that never occur?

Reply:
That will never occur. The authentication tag is either in the Data field, or in the `AuthenticatedEncryptionTag`.

The AuthenticatedEncryptionTag field does not exist in the protocol version below 1.4. So depending on the version, the tag is either already appeneded at the end of the cipher text (and the `AuthenticatedEncryptionTag` is empty) or it's in the `AuthenticatedEncryptionTag` only.

The authentication tag location prior to 1.4 however is not specified in kmip spec. It's just how most of servers have just implemented AES-GCM, when they implemented it.

One of the issue with the KMIP specifications is that many areas are left to the reader's free interpretation.

Reply:
So when I say "It will never occure", in fact I'm not 100% sure, but if it happens, the decryption request will be consistent with the encryption response anyway, and I would expect the server to properly accept such request.